src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala

```diff
@@ -77,6 +82,130 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
+    }
+  }
+
+  /**
+    * Currently just a copy from private method in basic trait! Will be optimized in future likely.
+    *
+    * Parse modifiers using specified serializer, check that its id is equal to the declared one,
+    * penalize misbehaving peer for every incorrect modifier,
+    * call deliveryTracker.onReceive() for every correct modifier to update its status
+    *
+    * @return collection of parsed modifiers
+    */
+  def parseModifiers[M <: NodeViewModifier](modifiers: Map[ModifierId, Array[Byte]],
+                                            serializer: ScorexSerializer[M],
+                                            remote: ConnectedPeer): Iterable[M] = {
+    modifiers.flatMap { case (id, bytes) =>
+      serializer.parseBytesTry(bytes) match {
+        case Success(mod) if id == mod.id =>
+          Some(mod)
+        case _ =>
+          // Penalize peer and do nothing - it will be switched to correct state on CheckDelivery
+          penalizeMisbehavingPeer(remote)
+          log.warn(s"Failed to parse modifier with declared id ${encoder.encodeId(id)} from ${remote.toString}")
+          None
+      }
+    }
+  }
+
+  /**
+    * Currently just copy from private method in basic trait! Will be optimized in future likely.
+    *
+    * Get modifiers from remote peer, filter out spam modifiers and penalize peer for spam
+    *
+    * @return ids and bytes of modifiers that were requested by our node
+    */
+  private def processSpam(remote: ConnectedPeer,
+                          typeId: ModifierTypeId,
+                          modifiers: Map[ModifierId, Array[Byte]]): Map[ModifierId, Array[Byte]] = {
+    val (requested, spam) = modifiers.partition { case (id, _) =>
+      deliveryTracker.status(id) == Requested
+    }
+
+    if (spam.nonEmpty) {
+      log.info(s"Spam attempt: peer $remote has sent a non-requested modifiers of type $typeId with ids" +
+        s": ${spam.keys.map(encoder.encodeId)}")
+      penalizeSpammingPeer(remote)
+    }
+    requested
+  }
+
+  /**
+    * Currently just copy from private method in basic trait! Will be optimized in future likely.
+    *
+    * Move `pmod` to `Invalid` if it is permanently invalid, to `Received` otherwise
+    */
+  private def validateAndSetStatus(remote: ConnectedPeer, pmod: ErgoPersistentModifier): Boolean = {
+    historyReaderOpt match {
+      case Some(hr) =>
+        hr.applicableTry(pmod) match {
+          case Failure(e) if e.isInstanceOf[MalformedModifierError] =>
+            log.warn(s"Modifier ${pmod.encodedId} is permanently invalid", e)
+            deliveryTracker.setInvalid(pmod.id)
+            penalizeMisbehavingPeer(remote)
+            false
+          case _ =>
+            deliveryTracker.setReceived(pmod.id, remote)
+            true
+        }
+      case None =>
+        log.error("Got modifier while history reader is not ready")
+        deliveryTracker.setReceived(pmod.id, remote)
```

Comment:
So in this case it got into Received state without being checked by applicableTry? So the `received` set may contain mods which are not applicable.